spec/formatting.md

```diff
@@ -18,3 +18,212 @@ the local variable takes precedence.
+    must result in an Unknown Function error if done within a context that
+    does not provide for the function `:func` to be successfully resolved:
+
+    ```
+    {The value is {(horse) :func}.}
+    ```
+
+    ```
+    match {(horse) :func}
+    when 1 {The value is one.}
+    when * {The value is not one.}
+    ```
+
+- **Selection errors** occur when message selection fails.
+
+  - **Selector errors** are failures in the matching of a key to a specific selector.
+
+    For example, attempting to format either of the following messages
+    may result in a Selector error if done within a context that
+    uses a `:plural` selector function which requires its input to be numeric:
+
+    ```
+    match {(horse) :plural}
+    when 1 {The value is one.}
+    when * {The value is not one.}
+    ```
+
+    ```
+    let $sel = {(horse) :plural}
+    match {$sel}
+    when 1 {The value is one.}
+    when * {The value is not one.}
+    ```
+
+- **Formatting errors** occur during the formatting of a resolved value,
+  for example when encountering a value with an unsupported type
+  or an internally inconsistent set of options.
+
+  For example, attempting to format any of the following messages
+  may result in a Formatting error if done within a context that
+
+  1. provides for the variable reference `$user` to resolve to
+     an object `{ name: 'Kat', id: 1234 }`, and
+  2. uses a `:get` formatting function which requires its argument to be an object and
+     an option `field` to be provided with a string value,
+
+  ```
+  {Hello, {(horse) :get field=name}!}
+  ```
+
+  ```
+  {Hello, {$user :get}!}
+  ```
+
+  ```
+  let $id = {$user :get field=id}
+  {Hello, {$id :get field=name}!}
+  ```
+
+Syntax and Data Model errors must be emitted as soon as possible.
+
+During selection, an Expression handler must only emit Resolution and Selection errors.
+During formatting, an Expression handler must only emit Resolution and Formatting errors.
+
+In all cases, when encountering an error,
+a message formatter must provide some representation of the message.
+An informative error or errors must also be separately provided.
+When a message contains more than one error,
+or contains some error which leads to further errors,
+an implementation which does not emit all of the errors
+should prioritise Syntax and Data Model errors over others.
+
+When an error occurs in the resolution of an Expression or Markup Option,
+the Expression or Markup in question is processed as if the option were not defined.
+This may allow for the fallback handling described below to be avoided,
+though an error must still be emitted.
+
+When an error occurs within a Selector,
+the selector must not match any VariantKey other than the catch-all `*`
+and a Resolution or Selector error is emitted.
+
+## Fallback String Representations
+
+The formatted string representation of a message with a Syntax or Data Model error
```

Comment:
Would it be better for each error to define the fallback representation rather than trying to collect them here? For example "The fallback string representation of an unknown function error is ..." in the section on that type of error?

Reply:
I would prefer defining the fallback representation according to the Placeholder for which it applies, rather than the error type. This makes them more predictable and intentionally restricts implementations from trying to be too smart about the behaviour.

Consider for instance something like `{$user :get key=name opt=$var}`, which is intended to get the user name with some `opt` qualifier. If the `$var` may be controlled by an attacker, it may lead to the `:get` formatting to fail, which means that a fallback representation must be used. If an implementation could define some custom error for it, a "smart" solution could be to try and stringify `$user`, which could lead to some sensitive information leaking out in an unexpected manner. Instead, with the explicit definition here it must use the string `{$user}`, which is sufficiently opaque to avoid this failure mode.

Reply:
I think I meant something different? I meant that instead of defining fallbacks in a separate section of the spec which then refers back to syntax and data model error types, it would potentially better to define error behavior inline with each error type. 

I am not suggesting that we encourage/permit different implementations to make up their own fallback behavior or to define new error types with (suspect) fallback behavior. 

One reason for my thinking is the example in the comment above:

```
let $var = {(horse) :unknown_function}
{You have a {$var}}
```

If I'm writing an implementation, I probably have an assignment operator bit of code for the `let (x) = {expr}`. One likely way to implement that would be to make the argument list a Map keyed with strings, into which the key `var` is put. When I do this, I'm not evaluating the rest of the pattern--to know how or even if the value of `var` is used. But I do need to assign `var` a value in the `let` expression. Defining the output of the "unknown function" error gives me clear guidance. Does that make sense? (This also helps answer @stasm's question about multiple errors)